src/gap_to_julia.jl

```diff
@@ -78,164 +94,153 @@ function gap_to_julia(::Type{Rational{T}}, x::Int64) where T <: Integer
 end
 
 function gap_to_julia(::Type{Rational{T}}, x::GapObj) where T <: Integer
-    if Globals.IsInt(x)
-        return gap_to_julia(T,x) // T(1)
-    end
-    if ! Globals.IsRat(x)
-        throw(ArgumentError("obj is not a rational"))
-    end
+    Globals.IsInt(x) && return gap_to_julia(T, x) // T(1)
+    ! Globals.IsRat(x) && throw(ConversionError(x, Rational{T}))
     numer = Globals.NumeratorRat(x)
     denom = Globals.DenominatorRat(x)
-    return gap_to_julia(T,numer) // gap_to_julia(T,denom)
+    return gap_to_julia(T, numer) // gap_to_julia(T, denom)
 end
 
 ## Floats
-function gap_to_julia( ::Type{Float64}, obj::GapObj)
-    if ! Globals.IsFloat(obj)
-        throw(ArgumentError("<obj> is not a MacFloat"))
-    end
-    return ValueMacFloat(obj)
+function gap_to_julia(::Type{Float64}, obj::GapObj)
+    Globals.IsIEEE754FloatRep(obj) && return ValueMacFloat(obj)::Float64
+    throw(ConversionError(obj, Float64))
 end
 
-gap_to_julia( ::Type{Float32}, obj::GapObj)  = Float32(gap_to_julia(Float64,obj))
-gap_to_julia( ::Type{Float16}, obj::GapObj)  = Float16(gap_to_julia(Float64,obj))
-gap_to_julia( ::Type{BigFloat}, obj::GapObj) = BigFloat(gap_to_julia(Float64,obj))
+gap_to_julia(::Type{T}, obj::GapObj) where T<:AbstractFloat = T(gap_to_julia(Float64, obj))
 
 ## Chars
 function gap_to_julia( ::Type{Char}, obj::GapObj)
-    if ! Globals.IsChar( obj )
-        throw(ArgumentError("argument is not a character object"))
-    end
-    return Char( Globals.INT_CHAR(obj ) )
+    Globals.IsChar(obj) && return Char(Globals.INT_CHAR(obj))
+    throw(ConversionError(obj, Char))
 end
 
 function gap_to_julia( ::Type{Cuchar}, obj::GapObj)
-    if ! Globals.IsChar( obj )
-        throw(ArgumentError("argument is not a character object"))
-    end
-    return trunc( Cuchar, Globals.INT_CHAR(obj ) )
+    Globals.IsChar(obj) && return trunc(Cuchar, Globals.INT_CHAR(obj))
+    throw(ConversionError(obj, Cuchar))
 end
 
-## Strings and symbols
-function gap_to_julia(::Type{String},obj::GapObj)
-    if Globals.IsStringRep(obj)
-        return CSTR_STRING(obj)
-    elseif Globals.IsString(obj)
-        return CSTR_STRING( Globals.CopyToStringRep( obj ) )
-    else
-        throw(ArgumentError("<obj> is not a string"))
-    end
+## Strings
+function gap_to_julia(::Type{String}, obj::GapObj)
+    Globals.IsStringRep(obj) && return CSTR_STRING(obj)
```

Comment:
what does `UNSAFE_CSTR_STRING` return? if there is a way to expose `Ptr{UInt8}` (i.e. pointer to raw memory), maybe `unsafe_string` would be better. maybe.

Reply:
I am not sure I understand what you are getting at here. `CSTR_STRING` is a helper function defined in GAP.jl which takes a GAP `T_STRING`, gets a `Ptr{UInt8}`, and converts it to a Julia string via `deepcopy(unsafe_string(ptr))`. And `UNSAFE_CSTR_STRING` is a badly-named sibling to that function which instead uses `unsafe_wrap(Array{UInt8,1}, char_ptr, string_len)`

That said,  the only place calling `UNSAFE_CSTR_STRING` immediately calls `deepcopy` on the result, too; I'd suggest that we move that `deepcoopy` into `UNSAFE_CSTR_STRING` and rename it to e.g. `CSTR_STRING_AS_ARRAY` or so

Reply:
first of all, the comment was probably misplaced ;) `UNSAFE_CSTR_STRING` is called a few lines lower;

but since we're at this here, according to docs there is no need for `deepcopy` after `unsafe_string` as it copies the data; it's `unsafe_wrap` that does not copy; 

Since you're discussing renaming: is there a reason of calling it `CSTR_STRING` instead of `unsafe_string`? (the same for `UNSAFE_CSTR_STRING`/`unsafe_wrap`)? (I assume `T_STRING` is a type? I couldn't find it through Github search).

Reply:
yes, `T_STRING` is the internal name inside the GAP kernel for the string type.

Thanks for the pointer on deepcopy. The names of the functions most likely have no deeper meaning; they simply are based on the name of the GAP kernel function `CSTR_STRING`. I don't mind renaming them at all.